### DISCUSSION TITLE
Native: 3D Terrain Support

### DISCUSSION THREAD

**Discussion by sjg-wdw** [maplibre/maplibre#326] [Native feature proposals]
### 3D Terrain Support

Here's some functionality we'd like to add to MapLibre. This is one proposal of several for Native and if you'd like a little context and our strategy [there's an overview](https://www.wetdogweather.com/2023/10/27/guide-to-maplibre-proposals/). Short version: Native first because we know it, then circle back and propose web.

This includes support for 3D terrain as implemented in the Javascript version, but goes beyond that to take a few more factors into account.

<img width="380" alt="Awesome new MapLibreJS 3D Terrain" src="https://github.com/maplibre/maplibre/assets/101820560/43c21d2d-0f6c-4461-852b-d2e66ad671aa">

## Visual 3D Terrain

At its most basic, the system asks for an elevation tile and then generates geometry with those elevations for the visual display.  Assuming you're in the same coordinate system for both display and data, this part isn't too bad.  That's what the Javascript version appears to be doing, and it's a fine approach.

We'd support this mode to start, which expects elevation for an entire source.  We'd also support a site mode, which would be focused on smaller areas for more constrained higher resolution data sets.

We would also want to add terrain animation as the data loads.  This would consist of elevation animating from zero to the target elevation.  That solves a tricky visual problem, discussed below.

## Elevation Data Sources

There are a lot of ways to represent elevation data.  The most compatible with MapLibre as it stands is individual tiles from a pyramid that stuff elevation into one or two channels of a standard image.  That's fine and should be the default implementation, but more flexibility is a good idea.

As such, we'd define an elevation loading interface that asked for data for a specific level and area.  Developers could then fill in their own data sources implementations which answered those requests in specific ways.  

For site mode you might simply have a flat elevation image of very high resolution.  By telling the system you only have one level, it would only ask for tiles within that level, showing the user as much as it could. For the more comprehensive world mode we might simply require lower levels of detail be made available.

There are also issues around the borders of tiles.  At the least we need on extra row and column at the edges to show polygons between the tiles.  For proper normals and shading we want at least two.  Most data formats don't provide this.  The solution is to make it optional, tracking data between tiles as it loads.

## Gesture Support

A flat map has its own set of gestures.  Add in a bit of tilt and you need to adjust those some.  Now add in terrain and things start to get strange.

Let's grab the top of this mountain.  Mostly works fine.

<img width="350" alt="Grab the mountain" src="https://github.com/maplibre/maplibre/assets/101820560/8f4497fb-19d2-42e9-a41b-fd8f5349bf16">

Now we pull it closer to us, shifting a bit to the left.  But the mouse is still pretty close to the middle.

<img width="350" alt="Grab the mountain" src="https://github.com/maplibre/maplibre/assets/101820560/c566f01d-d51b-40d5-86e6-bf41fc019aa1">

So what happened?  Without looking at the math I can only guess, but I suspect we grabbed the ground plane.  Unfortunately, doing it better is a lot of work and the library isn't really helping you out.

You need to figure out what is underneath the finger or cursor where they touch or tap.  You can get that from the depth buffer in the realtime rendering SDK, but it'll require a bit of work.  This approach would also be a good start for 3D buildings.

From experience I can say that no one is ever satisfied with your 3D gestures, so you want to make sure developers can write their own or enhance yours as needed.

## Terrain Following & Query

Now that the terrain is poking up in spots, your viewpoint can wind up underneath it.  This looks terrible and breaks the illusion, but with data loading unpredictably, it's going to happen and it's really a matter of how you deal with it.

<img width="350" alt="Behind the curtain" src="https://github.com/maplibre/maplibre/assets/101820560/ec3f9e72-7b68-4140-84cf-3bc5f3a0c0fd">

The approach we recommend is keeping the viewpoint above the terrain.  As terrain loads, give it an animated time and then push the user up above it to maintain the illusion.  You also need to mess with the near clipping plane to avoid just clipping the terrain.

I don't mean to pick on the MapLibreJS implementation.  It's good, but none of this is easy and the toolkit isn't making it any easier.  You really need to get deep in there to fix this stuff.

Beyond just the user's viewpoint, we'll want a way to query the elevation for random points and possibly linear features as well.  Polygons would be nice too, but get a bit complex.

If you want to decorate the terrain with features, you might want a way to 'subscribe' to elevation changes as new elevation data loads and old data unloads.

## Data Loading

The basic data loading will work initially, but has some pretty serious flaws where terrain is concerned.  

It turns out you can see mountains from far away, so only loading a single level of detail will get old quick.  What you really need is a combination of levels all at once.  Lower levels farther away and higher levels closer up.  You'd want a relatively seamless way to animate between those as you load new data and unload old data.

*We have a more comprehensive proposal to update the data loading for a globe and coordinate system support.  Adding elevation would require just one set of tweaks.*

Elevation makes tiles take up more space on the screen.  Rather than taking a flat rectangle and projecting it to the screen, you need a 3D box, with the maximum height in the tile taken into account.  We don't really know what that height is with the information we typically have in loading.

One solution is to give everything the height of Mt. Everest and call it a day.  A more nuanced approach is to track the maximum height of what you've already loaded.

## Vector Features

Sprinkling vector features (for points) or draping them (for lines, possibly polygons) on top of a terrain tile is an interesting problem.  We'd suggest starting with the simplest approach and keeping it open enough for others to get creative.

The simple approach is require all the parts of a tile to show up before doing anything.  In this case that would be the elevation, the vector features, and an image, if using one.  At that point we can nail the geometry down to the tile elevation all at once.

That means looking up elevation for the points and cutting lines wherever they intersect a terrain triangle.  If doing polygons, it's more complicated, but the same idea.  If there's an animation to get the geometry where it needs to go, then that's baked in at creation time.

For vector tiles, things get wild when combining low and high res tiles in the same view.  Let's ignore that for the moment.

## Discussion

There are a number of systems that work as described for data loading and you can tell that they don't work perfectly.  You'll find roads that flicker as they fight with the terrain, points that don't quite land right and features that end up fighting with each other.  All of that can be fixed, often by constraining the input data.  Actually, it's usually constraining the input data.

This proposal will get closer functionality-wise, but if you want a perfectly immersive earth view, it needs a really good database.

**Comment by HarelM** [comment]
Feel free to poke around the 3D terrain label for closed and open issue in maplibre gl js, I'm guessing you'll find some of the issues described here, some more, and possible missing issues.
You are also welcome to comment on some fixes that were made there and let us know if these fixes makes sense, for example this one: https://github.com/maplibre/maplibre-gl-js/pull/2759

**Reply by sjg-wdw** [reply]
That's a good one.  We have a pitch delegate in WhirlyGlobe that does this kind of thing.  It's a good idea.

**Comment by techmavengeospatial** [comment]
Supporting 3D Terrain from dynamic mesh created from RGB encoded PNG (heightmap tiles)

- MapZen Terrarium PNG  - inside MBTILES or GPKG and onliine
- Mapbox Terrain-RBG PNG  - inside MBTILES or GPKG  and onliine
- MapTiler Data RGB PNG  - inside MBTILES or GPKG  and onliine
- GDAL Tiled Gridded Coverage PNG GPKG http://www.geopackage.org/guidance/extensions/tiled_gridded_coverage_data.html

**Comment by NathanMOlson** [comment]
@sjg-wdw The link to the overview is broken: https://www.wetdogweather.com/blog/maplibre-proposals. Is that still available? Do you have any more technical details for the design ready to go somewhere?

**Reply by sjg-wdw** [reply]
New blog system.  The relevant discussions are here:
https://github.com/maplibre/maplibre/discussions/329
https://github.com/maplibre/maplibre/discussions/320
https://github.com/maplibre/maplibre/discussions/333
https://github.com/maplibre/maplibre/discussions/322

Though none of these is quite what Hudhud is doing.   So perhaps we should revisit that.

**Reply by HarelM** [reply]
Relevant initial implementation in web: 
- https://github.com/maplibre/maplibre-gl-js/pull/1022

There were a few other bug fixes later on. You can look at the `terrain` label in maplibre-gl-js repo.

**Reply by louwers** [reply]
Now seems to live under this URL https://www.wetdogweather.com/2023/10/27/guide-to-maplibre-proposals/

Updated the link in the first post.

**Comment by akrizs** [comment]
Is this something that is expected to be implemented into maplibre native, what method is recommended to get 3d terrain in native if its not available in maplibre?

**Reply by NathanMOlson** [reply]
@JesseCrocker started a branch to implement this, and I've moved it forward [here](https://github.com/NathanMOlson/maplibre-native/pull/1). At the top of the PR, I've got a (long) checklist of things that need to be done to finish this feature. I'm not actively working on it at the moment.

I think the best approach to get 3D terrain is to work on the TODO items for this branch, or pay someone else to.

@louwers Do you have any suggestions on how to collaborate across users on this large feature? Is it possible to set up a branch in the official repo with write permissions for multiple users who don't have write permission to `main`?

**Reply by louwers** [reply]
Hi @NathanMOlson, I just bumped you to write access.

Feel free to push it to a branch! Maybe `feature/terrain-3d`? For collaborating I still recommend creating PRs.

**Reply by NathanMOlson** [reply]
Done! Here it is: https://github.com/maplibre/maplibre-native/pull/4190

**Reply by louwers** [reply]
Thanks!
